### PR TITLE
Creating a simple language editor

### PR DESCRIPTION
**Pull request by etcho**
Creating a way to easily translate the files. When translating, it was really hard to know which strings were untranslated.

If it didn't fit in the project standards, understandable. I created it to help me to improve Portuguese translation and thought it might be useful.

### CODE REVIEW

**Comment by thorsten** [comment]
Thank you very much, we had an interface for translations in the v2.x versions but we removed it for v3.0 as it was a hell to maintain, unsecure and it had a very bad UI/UX.

I like the idea of having it as a side project for phpMyFAQ.

File: scripts/langEditor.php [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+ * It makes visible the strings already translated, the ones that exist on 
+ * a file but not on the original, and the ones that are not translated yet.
+ *
+ * @package   phpMyFAQ
+ * @author    Everton Leite <etcholeite@gmail.com>
+ */
+	
+?>
+
+<script src="https://code.jquery.com/jquery-3.6.0.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>
```
Comment (thorsten) [inline scripts/langEditor.php:14]: jQuery is already bundled in phpMyFAQ, so it can be used here.

File: scripts/langEditor.php [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	
+	.warning {
+		color: red;
+		margin: 10px 0px;
+	}
+</style>
+
+<?php
+
+$PMF_LANG_source = array();
```
Comment (thorsten) [inline scripts/langEditor.php:62]: You could use short arrays, so instead of `array()` you can use `[]`

File: scripts/langEditor.php [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+$LANG_CONF_source = array();
+$PMF_LANG_target = array();
+$LANG_CONF_target = array();
+$index_source = array();
```
Comment (thorsten) [inline scripts/langEditor.php:66]: Usually we use camelCase for variable names for phpMyFAQ.

File: scripts/langEditor.php [hunk 4]
```diff
@@ -0,0 +1,222 @@
+<?php
+	
+/**
+ * This file is intended to make it easier to translate the language files. 
+ * It makes visible the strings already translated, the ones that exist on 
+ * a file but not on the original, and the ones that are not translated yet.
+ *
```
Comment (thorsten) [inline scripts/langEditor.php:7]: Is it possible to use the MPL 2.0 here?

**Comment by etcho** [comment]
I was thinking about a standalone script, that's the reason for include jquery from the CDN. Now that you suggested making it part of the file tree, I will load the files automatically from the lang directory. Is EN the main language?

**Comment by thorsten** [comment]
Ah, you are right, a standalone script is a better solution.

English is the main language, right.

**Comment by etcho** [comment]
I have a doubt: should I always trim the strings? Sometimes there are mistakes solved with trim but sometimes looks like the spaces were put there on purpose.

File: scripts/langEditor.php [hunk 5]
```diff
@@ -16,7 +16,20 @@
  * @link      https://www.phpmyfaq.de
  * @since     2021-08-11
  */
-	
+
+//checking for sintax errors to prevent crashes when exporting the file to the application
+if ($_POST && key_exists("sintaxCheck", $_POST) && $_POST["sintaxCheck"] == 1 && key_exists("output", $_POST)) {
```
Comment (thorsten) [inline scripts/langEditor.php:21]: Please use single quotes instead of double quotes.

Comment (thorsten) [inline scripts/langEditor.php:21]: There's a typo: "syntax" instead of "sintax"

File: scripts/langEditor.php [hunk 6]
```diff
@@ -184,9 +219,14 @@ function renderSourceTableLine($index, &$pmfLangSource, &$pmfLangTarget, &$langC
...
-		<textarea id="output" placeholder="The content of your translated file will appear here"></textarea>
+		<div id="sintax_check_message"></div>
+		<textarea readonly="readonly" id="output" placeholder="The content of your translated file will appear here"></textarea>
 		
 		<script>
+			function copyFromSource(btn) {
+				btn.parentElement.nextElementSibling.querySelector(".target").value = btn.parentElement.previousElementSibling.innerHTML;
+			}
+		
 			var ready = (callback) => {
```
Comment (thorsten) [inline scripts/langEditor.php:230]: Please use let/const instead of var - we can use modern JavaScript. :)

File: scripts/langEditor.php [hunk 7]
```diff
@@ -212,6 +254,26 @@ function renderSourceTableLine($index, &$pmfLangSource, &$pmfLangTarget, &$langC
 						output += key + " = " + value + ";\n";
 					})
 					textarea.value = output;
+
+					//sending an ajax request with the output to check for possible sintax errors and typos
+					var xhttp = new XMLHttpRequest();
```
Comment (thorsten) [inline scripts/langEditor.php:259]: We can use the Fetch API here instead of XMLHttpRequest. See https://developer.mozilla.org/de/docs/Web/API/Fetch_API

Comment (etcho) [reply scripts/langEditor.php]: Thanks for the patience. I feel stuck on time, using jurassic js.

**Pull request merged**